src/net/tcp.rs

```diff
@@ -7,52 +7,76 @@ use std::io::{BufReader, BufWriter};
+type MessageReceiver = Option<Receiver<String>>;
 
-/// Wrapper around a TcpListener
-pub struct TcpServer {
-    connections: ConnectionMap,
-    // Control channel used to send messages to the *server* itself, *not* a specific client. Shutdown might be an example.
-    rx: MessageReceiver,
-    // Channel used for the server to send messages back up to the application
-    tx: MessageSender,
-    addr: SocketAddr,
+/// Container struct that keeps the hash map of connections
+pub struct TcpSocketState {
+    connections: Connections,
 }
 
-impl TcpServer {
-    /// Creates a new TCP server
-    pub fn new(addr: SocketAddr) ->  Result<TcpServer, io::Error> {
-        let tcp_server = TcpServer {
-            connections: Arc::new(RwLock::new(HashMap::new())),
-            rx: None,
-            tx: None,
-            addr: addr,
-        };
-        Ok(tcp_server)
+impl TcpSocketState {
+    /// Creates and returns a new TcpSocketState
+    pub fn new(addr: SocketAddr) -> TcpSocketState {
+        TcpSocketState{
+            connections: Arc::new(Mutex::new(HashMap::new()))
+        }
     }
 
+    /// This starts a TCP server on the provided SocketAddr. It is important to note that it also passes an Arc reference down to the server.
+    pub fn start(&mut self, addr: SocketAddr) {
+        TcpServer::listen(addr, self.connections.clone());
+    }
+}
+
+/// Wrapper around a TcpListener
+pub struct TcpServer;
+
+/// Using `self` to do deal with the threading proved to be very complicated. That is why these functions do use `self`.
+impl TcpServer {
+
     /// Starts the TcpServer listening socket. When a new connection is accepted, it spawns a new thread dedicated to that client and goes back to listening for more connections.
-    pub fn listen(&mut self, addr: SocketAddr) {
-        // TODO: Remove unwraps once we figure out error handling
-        let (tx, rx) = channel();
-        self.tx = Some(Box::new(tx));
-        self.rx = Some(Box::new(rx));
+    pub fn listen(addr: SocketAddr, connections: Connections) {
         thread::spawn(move || {
             let listener = TcpListener::bind(addr).unwrap();
+            /// This is a blocking call, so the thread waits here until it gets a new connection
             for stream in listener.incoming() {
-                thread::spawn(|| {
-                    let tcp_client = TcpClient::new(stream.unwrap());
-                });
+                /// Now we call a function and pass it the stream, and a clone of the connections hash
+                TcpServer::handle_connection(stream.unwrap(), connections.clone());
             }
         });
     }
+
+    /// This function inserts a reference to the connection into the connections hash
+    pub fn handle_connection(stream: TcpStream, connections: Connections) {
+        let peer_addr = stream.peer_addr().unwrap();
+        let mut tcp_client = Arc::new(Mutex::new(TcpClient::new(stream)));
+        if let Ok(mut lock) = connections.lock() {
```

Comment:
What happens if the lock is poisoned? in this case it would continue to work without adding the connection to the hashmap.